File: src/Java/Store/5. Final Hole/src/main/refactoringgolf/store/OrderItem.java
package refactoringgolf.store;

public class OrderItem {
	
	public Product product;
	public int quantity;

	public OrderItem(Product product, int quantity) {
		this.product = product;
		this.quantity = quantity;
	}
	
	public Product getProduct() {
		return product;
	}

	public int getQuantity() {
		return quantity;
	}

	public float total() {
		float unitPricePerQuanity = getUnitPricePerQuantity();
		float discount = createCategoryDiscount().calculateDiscount(this);
		return unitPricePerQuanity - discount;
	}

	private CategoryDiscount createCategoryDiscount() {
		CategoryDiscount categoryDiscount=null;
		if (getProduct().getCategory() == ProductCategory.Accessories) {
			categoryDiscount = new AccessoriesDiscount();
		}
	
		if (getProduct().getCategory() == ProductCategory.Components) {
			categoryDiscount = new ComponentsDiscount();
		}
	
		if (getProduct().getCategory() == ProductCategory.Bikes) {
			categoryDiscount = new BikesDiscount();
		}
	
		if (getProduct().getCategory() == ProductCategory.Cloathing) {
			categoryDiscount = new CloathingDiscount();
		}
		return categoryDiscount;
	}

	public float getUnitPricePerQuantity() {
		return getProduct().getUnitPrice() * getQuantity();
	}
}

File: src/Net/Store/After/OrderItem.cs
namespace After
{
    public class OrderItem
    {
        public Product Product { get; private set; }

        public int Quantity { get; private set; }

        public OrderItem(Product product, int quantity)
        {
            Product = product;
            Quantity = quantity;
        }

        public decimal CalculatePartialTotal()
        {
            return Product.UnitPrice * Quantity;
        }

        public decimal CalculateTotal()
        {
            decimal partialTotal = this.CalculatePartialTotal();
            decimal discount = this.CreateCategoryDiscount().CalculateDiscount(this);

            return partialTotal - discount;
        }

        private ICategoryDiscount CreateCategoryDiscount()
        {
            ICategoryDiscount categoryDiscount = null;
            if (this.Product.Category == ProductCategory.Accessories)
            {
                categoryDiscount = new AccessoriesDiscount();
            }

            if (this.Product.Category == ProductCategory.Components)
            {
                categoryDiscount = new ComponentsDiscount();
            }

            if (this.Product.Category == ProductCategory.Bikes)
            {
                categoryDiscount = new BikesDiscount();
            }

            if (this.Product.Category == ProductCategory.Cloathing)
            {
                categoryDiscount = new CloathingDiscount();
            }
            return categoryDiscount;
        }
    }
}